File: examples/nodejs/producer.js
#!/usr/bin/env node
/**
 * Kafka Producer Example (Node.js)
 * 
 * This script sends messages to a Kafka topic using KafkaJS.
 */

const { Kafka } = require('kafkajs');

// Configuration
const BOOTSTRAP_SERVERS = ['kafka-kafka-bootstrap.kafka.svc:9092']; // Internal access
// const BOOTSTRAP_SERVERS = ['<your-loadbalancer-endpoint>:9094']; // External access
const TOPIC = 'demo-topic';
const CLIENT_ID = 'nodejs-producer';

// Create Kafka client
const kafka = new Kafka({
  clientId: CLIENT_ID,
  brokers: BOOTSTRAP_SERVERS,
  retry: {
    retries: 5,
    initialRetryTime: 300,
  },
});

// Create producer
const producer = kafka.producer({
  allowAutoTopicCreation: false,
  transactionalId: undefined,
  maxInFlightRequests: 5,
  idempotent: true,
  acks: -1, // Wait for all replicas
});

async function sendMessage(message) {
  try {
    const result = await producer.send({
      topic: TOPIC,
      messages: [
        {
          key: message.message_id.toString(),
          value: JSON.stringify(message),
          headers: {
            'content-type': 'application/json',
          },
        },
      ],
    });

    console.log(`✓ Message sent to partition ${result[0].partition} at offset ${result[0].baseOffset}`);
    return true;
  } catch (error) {
    console.error(`✗ Failed to send message: ${error.message}`);
    return false;
  }
}

async function main() {
  try {
    // Connect producer
    console.log('🔄 Connecting to Kafka...');
    await producer.connect();
    console.log(`✓ Connected to Kafka at ${BOOTSTRAP_SERVERS.join(', ')}`);

    // Send 10 messages
    for (let i = 0; i < 10; i++) {
      const message = {
        message_id: i,
        timestamp: new Date().toISOString(),
        content: `Hello Kafka! Message #${i}`,
        metadata: {
          source: 'nodejs-producer',
          version: '1.0',
        },
      };

      console.log(`\n📤 Sending message ${i}...`);
      if (await sendMessage(message)) {
        console.log(`   Data: ${message.content}`);
      }

      // Wait 1 second between messages
      await new Promise(resolve => setTimeout(resolve, 1000));
    }

    console.log(`\n✓ Successfully sent 10 messages to topic '${TOPIC}'`);

  } catch (error) {
    console.error(`\n✗ Error: ${error.message}`);
    process.exit(1);
  } finally {
    // Disconnect producer
    console.log('\n🔄 Disconnecting producer...');
    await producer.disconnect();
    console.log('✓ Producer disconnected');
  }
}

// Handle interruption
process.on('SIGINT', async () => {
  console.log('\n⚠ Interrupted by user');
  await producer.disconnect();
  process.exit(0);
});

// Run main function
main().catch(console.error);

File: examples/python/producer.py
#!/usr/bin/env python3
"""
Kafka Producer Example

This script sends messages to a Kafka topic.
"""

import json
import time
from datetime import datetime
from kafka import KafkaProducer
from kafka.errors import KafkaError

# Configuration
BOOTSTRAP_SERVERS = ['kafka-kafka-bootstrap.kafka.svc:9092']  # Internal access
# BOOTSTRAP_SERVERS = ['<your-loadbalancer-endpoint>:9094']   # External access
TOPIC = 'demo-topic'

def create_producer():
    """Create and return a Kafka producer."""
    try:
        producer = KafkaProducer(
            bootstrap_servers=BOOTSTRAP_SERVERS,
            value_serializer=lambda v: json.dumps(v).encode('utf-8'),
            # Retry configuration
            retries=5,
            max_in_flight_requests_per_connection=1,
            # Acknowledgment configuration
            acks='all',  # Wait for all replicas to acknowledge
        )
        print(f"✓ Connected to Kafka at {BOOTSTRAP_SERVERS}")
        return producer
    except KafkaError as e:
        print(f"✗ Failed to create producer: {e}")
        raise

def send_message(producer, topic, message):
    """Send a message to Kafka topic."""
    try:
        # Send message asynchronously
        future = producer.send(topic, value=message)
        
        # Block until message is sent (or timeout)
        record_metadata = future.get(timeout=10)
        
        print(f"✓ Message sent to partition {record_metadata.partition} "
              f"at offset {record_metadata.offset}")
        return True
    except KafkaError as e:
        print(f"✗ Failed to send message: {e}")
        return False

def main():
    """Main function to send messages."""
    producer = create_producer()
    
    try:
        # Send 10 messages
        for i in range(10):
            message = {
                'message_id': i,
                'timestamp': datetime.now().isoformat(),
                'content': f'Hello Kafka! Message #{i}',
                'metadata': {
                    'source': 'python-producer',
                    'version': '1.0'
                }
            }
            
            print(f"\n📤 Sending message {i}...")
            if send_message(producer, TOPIC, message):
                print(f"   Data: {message['content']}")
            
            # Wait 1 second between messages
            time.sleep(1)
        
        print(f"\n✓ Successfully sent 10 messages to topic '{TOPIC}'")
        
    except KeyboardInterrupt:
        print("\n⚠ Interrupted by user")
    finally:
        # Flush and close producer
        print("\n🔄 Flushing remaining messages...")
        producer.flush()
        producer.close()
        print("✓ Producer closed")

if __name__ == '__main__':
    main()
